src/emmc.c

```diff
@@ -0,0 +1,190 @@
+}
+
+gboolean r_emmc_read_bootpart(const gchar *device, gint *bootpart_active, GError **error) {
+	int ret;
+	gboolean res = FALSE;
+	__u8 extcsd[512];
+	int fd;
+	/* count from 1 */
+	int active_partition = -1;
+
+	fd = open(device, O_RDONLY);
+
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening output device failed: %s", strerror(errno));
+		goto out;
+	}
+
+	ret = r_emmc_read_extcsd(fd, extcsd);
+
+	if (ret) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not read from extcsd register %d in %s\n",
+				EXT_CSD_PART_CONFIG, device);
+		goto out;
+	}
+
+	/* retrieve active partition from BOOT_PART register */
+	active_partition = (extcsd[EXT_CSD_PART_CONFIG] & 0x38)>>3;
+
+	/* return the partitions counted from 0 */
+	*bootpart_active = active_partition-1;
+
+	res = TRUE;
+
+out:
+	if (fd)
+		close(fd);
+
+	return res;
+}
+
+static int r_emmc_write_extcsd(int fd, __u8 index, __u8 value) {
+	struct mmc_ioc_cmd cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	cmd.write_flag = 1;
+	cmd.opcode = MMC_SWITCH;
+	cmd.arg = (MMC_SWITCH_MODE_WRITE_BYTE << 24) | (index << 16) |
+	          (value << 8) | EXT_CSD_CMD_SET_NORMAL;
+	cmd.flags = MMC_RSP_SPI_R1B | MMC_RSP_R1B | MMC_CMD_AC;
+
+	return ioctl(fd, MMC_IOC_CMD, &cmd);
+}
+
+gboolean r_emmc_write_bootpart(const gchar *device, gint bootpart_active, GError **error) {
+	int ret;
+	gboolean res = FALSE;
+	int fd;
+	__u8 value = 0;
+
+	g_assert(bootpart_active == 0 || bootpart_active == 1);
+
+	fd = open(device, O_RDWR);
+
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening output device failed: %s", strerror(errno));
+		goto out;
+	}
+
+	/* write [5:3] : BOOT_PARTITION_ENABLE of PARTITION_CONFIG */
+	if (bootpart_active == 0)
+		value = 0x08;
+	else if (bootpart_active == 1)
+		value = 0x10;
+
+	ret = r_emmc_write_extcsd(fd, EXT_CSD_PART_CONFIG, value);
+	if (ret) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not write 0x%02x to extcsd register %d in %s\n",
+				value, EXT_CSD_PART_CONFIG, device);
+		goto out;
+	}
+
+	res = TRUE;
+
+out:
+	if (fd)
+		close(fd);
+
+	return res;
+}
+
+static gboolean r_emmc_force_part_write(const gchar *device, gchar value, GError **error) {
+	gboolean ret = FALSE;
+	gchar **device_split = NULL;
+	const gchar *sysfs_path = NULL;
+	FILE *f = NULL;
+
+	g_assert(value == '0' || value == '1');
+
+	/* extract mmcblkXbootY from /dev/mmcblkXbootY */
+	device_split = g_strsplit(device, "/", 3);
+
+	if (g_strv_length(device_split) != 3) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not generate sysfs path from device: split failed");
+		goto out;
+	}
+
+	sysfs_path = g_strdup_printf("/sys/block/%s/force_ro", device_split[2]);
+
+	if (!g_file_test(sysfs_path, G_FILE_TEST_EXISTS)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not generate sysfs path from device: %s does not exist",
```

Comment:
The problem is not generating the path here, but the configured block devices doesn't have the force_ro attribute.